File: java/src/main/java/br/datastructures/queue/ArrayQueue.java
package br.datastructures.queue;

import br.datastructures.exceptions.NoSuchItemException;

public class ArrayQueue<I> extends AbstractQueue<I> {
    private int first, last;
    private I items[];
    private int capacity;
    
    public ArrayQueue(int initialSize) {
        first = -1;
        last = -1;
        capacity = initialSize;
        items = (I[]) new Object[capacity];
    }
    
    private boolean full() {
        return size() == capacity;
    }
    
    @Override
    public I getFirst() throws NoSuchItemException {
        if(empty()) 
            throw new NoSuchItemException();
        
        return items[first];
    }
    
    private I[] resize() {
        capacity *= 2;
        I newItems[] = (I[]) new Object[capacity];
        System.arraycopy(items, first, newItems, 0, size());
        
        return newItems;
    }
    
    @Override
    public void enqueue(I item) {
        if(full()) 
            items = resize();
        
        if (empty()) 
            first = 0;
        
        last++;
        items[last] = item;
        quantity++;
    }
    
    @Override
    public I dequeue() throws NoSuchItemException {
        I item = getFirst();
        first++;
        
        if(first > last)
            first = last = -1;
        
        quantity--;

        return item;
    }
}

File: c/queue/ArrayQueue.c
#include <stdio.h>
#include <stdlib.h>

#define NENHUM_ITEM -1

typedef int Item;

typedef struct {
    Item* itens;
    int capacity, quantity, first, last;
} ArrayQueue;

ArrayQueue* newArrayQueue(int capacity) {
    ArrayQueue* queue = (ArrayQueue*) malloc(sizeof(ArrayQueue));
    Item* itens = (Item*) malloc(capacity * sizeof(Item));

    queue->itens = itens;
    queue->capacity = capacity;
    queue->quantity = 0;
    queue->first = -1;
    queue->last = -1;

    return queue;
}

int size(ArrayQueue* queue) {
    return queue->quantity;
}

int empty(ArrayQueue* queue) {
    return size(queue) == 0;
}

ArrayQueue* resize(ArrayQueue* queue) {
    int newCapacity = queue->capacity * 2;
    Item* itens = realloc(queue->itens, newCapacity * sizeof(Item));

    if(itens == NULL)
        return NULL;

    queue->itens = itens;    
    queue->capacity = newCapacity;

    return queue;
}

int full(ArrayQueue* queue) {
    return queue->capacity == queue->quantity;
}

void enqueue(ArrayQueue* queue, Item item) {
    ArrayQueue* copiaQueue = queue;
    
    if(full(queue))
        copiaQueue = resize(queue);
    
    if(copiaQueue == NULL)
        return;

    queue = copiaQueue;

    if(empty(queue)) 
        queue->first = queue->last = 0;

    queue->itens[queue->last++] = item;

    queue->quantity++;
}

Item getFirst(ArrayQueue* queue) {
    if(empty(queue))
        return NENHUM_ITEM;

    return queue->itens[queue->first];
}

Item dequeue(ArrayQueue* queue) {
    if(empty(queue))
        return NENHUM_ITEM;

    Item first = getFirst(queue);
    queue->quantity--;
    queue->first++;

    if(empty(queue))
        queue->first = queue->last = -1;

    return first;
}

void viewQueue(ArrayQueue* queue) {
    printf("Itens da lista:");
    while(!empty(queue))
        printf(" %d", dequeue(queue));

    printf("\n");
}

void freeQueue(ArrayQueue* queue) {
    free(queue->itens);
    free(queue);
}

int main() {
    ArrayQueue* queue = newArrayQueue(5);

    enqueue(queue, 10);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 11);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 12);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 13);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 14);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 15);
    printf("Primeiro item: %d\n", getFirst(queue));

    printf("%d removido\n", dequeue(queue));

    viewQueue(queue);

    freeQueue(queue);

    return 0;
}